### ISSUE TITLE
Thoughts on Python package naming/renaming?

### ISSUE THREAD

**Issue by mikepurvis** [colcon/colcon-core#380]
Currently my org builds a number of plain Python packages in our catkin_tools workspace, using a fairly simple [catkin_tools_python](https://github.com/mikepurvis/catkin_tools_python) plugin which recognizes the `<build_type>python</build_type>` and does the right thing as far as calling through to the `setup.py` script. However, obviously for this to work we need to generate the appropriate `package.xml` files, and this means having to mirror Python repos (or even pypi source tarballs) internally, and patch them in. In order to be compatible with the similar translation that happens when pypi packages are released into Debian, part of this involves prefixing the names with `python-` or `python3-`. Currently, this process is very naive and doesn't try to utilize `rosdep` or any external mapping info— the handful of exception cases are just managed as a manual overrides list:

https://github.com/mikepurvis/catkin_tools_python/blob/85f746576ea501b30df991b5c00ab8ae0fb5364a/catkin_tools_python/filters.py#L15-L21

As we contemplate moving to colcon, being able to build native Python packages without patching in extra metadata is obviously a win. However, as currently set up, colcon's package discovery mechanism does not do the prefixing, for example from a checkout of the [requests](https://github.com/psf/requests) source:

```
requests {path: requests, type: python, name: requests, dependencies: defaultdict(<class 'set'>, {'build': set(), 'run': {'certifi', 'urllib3', 'idna', 'chardet'}, 'test': {'PySocks', 'pytest-httpbin', 'pytest-cov', 'pytest', 'pytest-mock', 'pytest-xdist'}}), hooks: [], metadata: {'get_python_setup_options': <function PythonPackageIdentification.identify.<locals>.getter at 0x7fa09ade2700>, 'version': '2.24.0'}}
```

This is a barrier to having a catkin/ament package be dependent on a Python package in a way that is agnostic to whether that Python package is in the workspace as source, or packaged.

For my immediate needs, I've built a [small augmentation plugin](https://github.com/mikepurvis/colcon-python-naming/blob/master/colcon_python_naming/__init__.py) which does the right thing:

```
$ colcon list
python3-requests {path: requests, type: python, name: python3-requests, dependencies: defaultdict(<class 'set'>, {'build': set(), 'run': {'python3-urllib3', 'python3-certifi', 'python3-chardet', 'python3-idna'}, 'test': {'python3-pytest-httpbin', 'python3-pytest-mock', 'python3-pytest', 'python3-pysocks', 'python3-pytest-xdist', 'python3-pytest-cov'}}), hooks: [], metadata: {'get_python_setup_options': <function PythonPackageIdentification.identify.<locals>.getter at 0x7fef671dc700>, 'version': '2.24.0'}}
python3-requests        requests        (python)
```

However, I'm wondering if this is something which colcon could/should deal with natively somehow, given the importance of being able to interoperate across the python and native package namespaces.

**Comment by dirk-thomas** [comment]
I understand the need to generate `package.xml` files when using `catkin_tools` but in the case of `colcon` that shouldn't be necessary. And the dependencies declared in the `setup.py` / `setup.cfg` file shouldn't have any kind of prefix (or even `rosdep` specific names). So I am wondering why you couldn't build the Python packages just as is (without a `package.xml` file)? This should at least make dependencies between such Python packages work.

> colcon's package discovery mechanism does not do the prefixing

That is correct. But I would argue that `colcon` shouldn't do that since this is a highly platform specific mangling.

> This is a barrier to having a catkin/ament package be dependent on a Python package in a way that is agnostic to whether that Python package is in the workspace as source, or packaged.
> I'm wondering if this is something which colcon could/should deal with natively somehow, given the importance of being able to interoperate across the python and native package namespaces.

I understand the problem for your use case. The actual source of the problem is that `rosdep` keys used in `package.xml` files (which are not ROS package names) are arbitrary.

Rather than letting `colcon` prepend arbitrary (potentially incorrect) prefixes to the package name I think the problem should be resolved at the source of the inconsistency: the `rosdep` keys. If `rosdep` could  provide the actual Python package name for such a key `colcon` could use that package name for the dependency (instead of using the rosdep key itself). That kind of mapping could also easily be implemented in `colcon-ros` where it conceptionally belongs.

Since rosdep keys only provide a mapping per platform (and not per package manager) we currently can't assume that every `rosdep` key for a Python package has a `pip` section which would state the actual package name. I am not sure if iterating all platforms and checking for `pip` information would be sufficient for this approach to work?

**Comment by mikepurvis** [comment]
Yeah, I definitely don't want to have to generate `package.xml` files for Python packages any more!

You're there, but just to be explicit, the situation is that I have a pile of packages, where some are catkin, with `<exec_depend>` dependencies on debian/rosdep names like `python3-requests`, and others are plain Python packages, which identify themselves and their own dependencies by the pypi name (like `requests`). So I need to be able to call `PackageDescriptor::get_recursive_dependencies` with the list of all these packages, and have them actually resolve (which they won't do today).

My hack approach is to augment the python packages so they identify themselves by the rosdep names, and dependency resolution works naturally (and it will interoperate gracefully in a transitional world where some still have `package.xml` files patched in). If I understand your proposal, it's to reverse this and basically augment the catkin/ament package DependencyDescriptor entities to map those names back to the original names for the purposes of dependency resolution.

Overall, this makes sense to me. From a purely pragmatic point of view, my inclination would be to use the rosdep database if available, but also have at least as an option a fallback which just mindlessly strips the `python-` and `python3-` prefix from dependencies, on the grounds that in the content of dependency resolution, this will basically never be wrong behaviour, and is much safer than the other approach of mindlessly prefixing. In particular, it would be nice if `rosdep` (and in particular the `rosdep init` step) could avoid becoming a hard dependency of `colcon-ros`.

**Comment by dirk-thomas** [comment]
> If I understand your proposal, it's to reverse this and basically augment the catkin/ament package DependencyDescriptor entities to map those names back to the original names for the purposes of dependency resolution.

Correct.

> From a purely pragmatic point of view, my inclination would be to use the rosdep database if available, but also have at least as an option a fallback which just mindlessly strips the `python-` and `python3-` prefix from dependencies, on the grounds that in the content of dependency resolution, this will basically never be wrong behaviour, and is much safer than the other approach of mindlessly prefixing.

Sounds reasonable to me. I would foresee case where this simply wouldn't work. E.g. a package manifest depending on the rosdep key `python(3)-catkin-pkg-modules`. In that case I think the only viable way would be to query rosdep and get the package name used for `pip`.
